### PR TITLE
Cake5

### PR DESCRIPTION
**Pull request by JustinRuiter**
An updated version of MolbioUnige's PR, to fix the pipeline and fix various typing issues.

PR into MolbioUnigue's repo: https://github.com/MolbioUnige/Webservice/pull/1

### CODE REVIEW

**Comment by ADmad** [comment]
As I mentioned on the related issue your PR needs to target the `cake5` branch not `master`, so please rebase your branch onto upstream ~~master~~ `cake5` and re-target the PR.

**Comment by ADmad** [comment]
You just changed the branch target without rebasing and fixing merge conflicts :slightly_smiling_face:

**Comment by ADmad** [comment]
Sorry, I made a typo in my previous comment, you need to rebase upon `cake5` not `master`.

**Comment by JustinRuiter** [comment]
> Sorry, I made a typo in my previous comment, you need to rebase upon `cake5` not `master`.

Rebased

File: phpunit.xml.dist [hunk 1]
```diff
@@ -1,23 +1,20 @@
...
-<phpunit
-    bootstrap="./tests/bootstrap.php"
-    colors="true"
-    stopOnFailure="false"
-    >
-
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" bootstrap="./tests/bootstrap.php" colors="true" stopOnFailure="false" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/10.2/phpunit.xsd" cacheDirectory=".phpunit.cache">
     <testsuites>
-        <testsuite name="Webservice Test Cases">
+        <testsuite name="Slug Test Cases">
```
Comment (ADmad) [inline phpunit.xml.dist]:
```suggestion
        <testsuite name="Webservice Test Cases">
```

File: src/Datasource/Connection.php [hunk 2]
```diff
@@ -33,17 +41,68 @@ class Connection
...
-        /** @psalm-var class-string<\Muffin\Webservice\Webservice\Driver\AbstractDriver> */
         $driver = $config['driver'];
         unset($config['driver'], $config['service']);
 
-        $this->_driver = new $driver($config);
+        $tempDriver = new $driver($config);
 
-        /** @psalm-suppress TypeDoesNotContainType */
-        if (!($this->_driver instanceof AbstractDriver)) {
+        if (!($tempDriver instanceof AbstractDriver)) {
```
Comment (ADmad) [inline src/Datasource/Connection.php]: We don't need this explicit check anymore since the property is typed now.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -14,11 +14,11 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        php-version: ['7.4', '8.0', '8.3']
+        php-version: ['8.2', '8.3']
         db-type: ['mysql', 'pgsql']
         prefer-lowest: ['']
         include:
-          - php-version: '7.4'
+          - php-version: '8.3'
```
Comment (ADmad) [inline .github/workflows/ci.yml]:
```suggestion
          - php-version: '8.1'
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -81,16 +81,17 @@ jobs:
...
-        tools: psalm:4, phpstan:1
+        tools: psalm:5.23.1, phpstan:1.10.65
 
     - name: Composer Install
       run: composer require cakephp/cakephp-codesniffer:^4.2
 
     - name: Run phpcs
-      run: vendor/bin/phpcs --standard=CakePHP src/ tests/
+      # Exclude Type hint sniffing, as it interferes with psalm
+      run: vendor/bin/phpcs --standard=CakePHP --exclude=CakePHP.Classes.ReturnTypeHint src/ tests/
```
Comment (ADmad) [inline .github/workflows/ci.yml]: Please don't exclude the sniff, you can add the psalm errors to it's baseline file instead so that they can be sorted out later.

File: composer.json [hunk 5]
```diff
@@ -39,12 +39,19 @@
...
-        "cakephp/orm": "^4.5"
+        "cakephp/orm": "^5.0"
     },
     "require-dev": {
-        "cakephp/cakephp": "^4.5",
-        "cakephp/cakephp-codesniffer": "^4.2",
-        "phpunit/phpunit": "^9.3"
+        "cakephp/cakephp": "^5.0",
+        "cakephp/cakephp-codesniffer": "^5.0",
+        "phpunit/phpunit": "^10.1",
```
Comment (ADmad) [inline composer.json]:
```suggestion
        "phpunit/phpunit": "^10.5.5",
```
This is what the core has.

File: src/Datasource/Connection.php [hunk 6]
```diff
@@ -33,17 +41,68 @@ class Connection
      */
     public function __construct(array $config)
     {
+        if (isset($config['name'])) {
+            $this->configName = $config['name'];
```
Comment (ADmad) [inline src/Datasource/Connection.php:44]: There's no need for a separate property. Can directly read the name from `$_config`.

File: src/Datasource/Connection.php [hunk 7]
```diff
@@ -33,17 +41,68 @@ class Connection
...
+    }
+
+    /**
+     * Get the configuration name for this connection.
+     *
+     * @return string
+     */
+    public function configName(): string
+    {
+        return $this->configName;
```
Comment (ADmad) [inline src/Datasource/Connection.php:96]:
```suggestion
        return $this->_config['name'] ?? '';
```

File: src/Datasource/Connection.php [hunk 8]
```diff
@@ -61,7 +120,7 @@ protected function _normalizeConfig(array $config): array
                 throw new MissingConnectionException(['name' => $config['name']]);
             }
 
-            $config['driver'] = App::className($config['service'], 'Webservice/Driver');
+            $config['driver'] = App::className($config['service'], 'Webservice/Driver', 'Driver');
```
Comment (ADmad) [inline src/Datasource/Connection.php]:
```suggestion
            $config['driver'] = App::className($config['service'], 'Webservice/Driver');
```

Currently drivers don't have the `Driver` suffix (similar to database driver classes in the core) and there's no need to change it.

File: src/Datasource/Marshaller.php [hunk 9]
```diff
@@ -110,25 +109,21 @@ protected function _validate(array $data, array $options, bool $isNew): array
         if (!$options['validate']) {
             return [];
         }
-
-        $validator = null;
         if ($options['validate'] === true) {
-            $validator = $this->_endpoint->getValidator('default');
```
Comment (ADmad) [inline src/Datasource/Marshaller.php]: Why change all these if / else blocks?

File: src/Datasource/Query.php [hunk 10]
```diff
@@ -99,12 +161,247 @@ public function __construct(WebserviceInterface $webservice, Endpoint $endpoint)
...
+     * Set the default repository object that will be used by this query.
+     *
+     * @param \Cake\Datasource\RepositoryInterface $repository The default repository object to use.
+     * @return $this
+     */
+    public function setRepository(RepositoryInterface $repository): Query
+    {
+        if ($repository instanceof Endpoint) {
+            $this->_endpoint = $repository;
+        }
```
Comment (ADmad) [inline src/Datasource/Query.php]:
```suggestion
        assert(
            $repository instanceof Endpoint,
            '`$repository` must be an instance of `' . Endpoint::class . '`.'
        );

        $this->_endpoint = $repository;
```

File: src/Datasource/Query.php [hunk 11]
```diff
@@ -99,12 +161,247 @@ public function __construct(WebserviceInterface $webservice, Endpoint $endpoint)
...
+        return $this;
+    }
+
+    /**
+     * Returns the default repository object that will be used by this query,
+     * that is, the table that will appear in the from clause.
+     *
+     * @return \Cake\Datasource\RepositoryInterface
+     */
+    public function getRepository(): RepositoryInterface
```
Comment (ADmad) [inline src/Datasource/Query.php]:
```suggestion
    public function getRepository(): Endpoint
```

File: src/Datasource/Query.php [hunk 12]
```diff
@@ -199,7 +495,7 @@ public function getEndpoint(): Endpoint
      * @param \Muffin\Webservice\Webservice\WebserviceInterface $webservice The webservice to use
      * @return $this
      */
-    public function setWebservice(WebserviceInterface $webservice)
+    public function setWebservice(WebserviceInterface $webservice): Query
```
Comment (ADmad) [inline src/Datasource/Query.php:391]:
```suggestion
    public function setWebservice(WebserviceInterface $webservice)
```

Functions returning `$this` shouldn't have a return type.

File: src/Datasource/Query.php [hunk 13]
```diff
@@ -211,40 +507,18 @@ public function setWebservice(WebserviceInterface $webservice)
      *
      * @return \Muffin\Webservice\Webservice\WebserviceInterface
      */
-    public function getWebservice()
+    public function getWebservice(): WebserviceInterface
     {
         return $this->_webservice;
     }
 
-    /**
```
Comment (ADmad) [inline src/Datasource/Query.php]: Please move back this function to it's original location and avoid unnecessary diff changes

File: src/Datasource/Query.php [hunk 14]
```diff
@@ -396,7 +666,7 @@ public function set($fields = null)
     /**
      * @inheritDoc
      */
-    public function offset($offset)
+    public function offset(?int $offset): Query|QueryInterface
```
Comment (ADmad) [inline src/Datasource/Query.php]: Remove return types from all functions that return `$this`.

File: src/Model/Endpoint.php [hunk 15]
```diff
@@ -136,7 +150,7 @@ class Endpoint implements RepositoryInterface, EventListenerInterface, EventDisp
      *
      * @param array $config List of options for this endpoint
      */
-    public function __construct(array $config = [])
+    final public function __construct(array $config = [])
```
Comment (ADmad) [inline src/Model/Endpoint.php]:
```suggestion
    public function __construct(array $config = [])
```

File: src/Model/Endpoint.php [hunk 16]
```diff
@@ -304,9 +330,9 @@ public function setConnection(Connection $connection)
     /**
      * Returns the connection driver.
      *
-     * @return \Muffin\Webservice\Datasource\Connection
+     * @return \Muffin\Webservice\Datasource\Connection|null
      */
-    public function getConnection(): Connection
+    public function getConnection(): ?Connection
```
Comment (ADmad) [inline src/Model/Endpoint.php]: This return shouldn't be nullable, throw an exception if the property is `null`.

File: src/Model/Endpoint.php [hunk 17]
```diff
@@ -542,24 +542,26 @@ public function getInflectionMethod(): string
...
     }
 
     /**
      * Get this endpoints associated webservice
      *
-     * @return \Muffin\Webservice\Webservice\WebserviceInterface
+     * @return \Muffin\Webservice\Webservice\WebserviceInterface|null
      */
-    public function getWebservice(): WebserviceInterface
+    public function getWebservice(): ?WebserviceInterface
```
Comment (ADmad) [inline src/Model/Endpoint.php]:
```suggestion
    public function getWebservice(): WebserviceInterface
```

File: src/Model/Endpoint.php [hunk 18]
```diff
@@ -659,6 +661,9 @@ public function findAll(Query $query, array $options): Query
      */
     public function findList(Query $query, array $options): Query
```
Comment (ADmad) [inline src/Model/Endpoint.php:674]: This method needs to be updated similar to core's `Table::findList()`

File: src/Model/EndpointLocator.php [hunk 19]
```diff
@@ -35,12 +36,11 @@ public function set(string $alias, RepositoryInterface $repository): Endpoint
      *
      * @param string $alias The alias name you want to get.
      * @param array $options The options you want to build the endpoint with.
-     * @return \Muffin\Webservice\Model\Endpoint
+     * @return \Cake\Datasource\RepositoryInterface
      * @throws \RuntimeException If the registry alias is already in use.
      */
-    public function get(string $alias, array $options = []): Endpoint
+    public function get(string $alias, array $options = []): RepositoryInterface
```
Comment (ADmad) [inline src/Model/EndpointLocator.php]:
```suggestion
    public function get(string $alias, array $options = []): Endpoint
```

Endpoint locator must always return an Endpoint :)

File: src/Model/EndpointLocator.php [hunk 20]
```diff
@@ -49,9 +49,9 @@ public function get(string $alias, array $options = []): Endpoint
      *
      * @param string $alias Endpoint alias.
      * @param array $options The alias to check for.
-     * @return \Muffin\Webservice\Model\Endpoint
+     * @return \Cake\Datasource\RepositoryInterface
      */
-    protected function createInstance(string $alias, array $options)
+    protected function createInstance(string $alias, array $options): RepositoryInterface
```
Comment (ADmad) [inline src/Model/EndpointLocator.php]:
```suggestion
    protected function createInstance(string $alias, array $options): Endpoint
```

File: src/Model/Resource.php [hunk 21]
```diff
@@ -30,7 +30,7 @@ class Resource implements EntityInterface, InvalidPropertyInterface
      * @param array $properties hash of properties to set in this resource
      * @param array $options list of options to use when creating this resource
      */
-    public function __construct(array $properties = [], array $options = [])
+    final public function __construct(array $properties = [], array $options = [])
```
Comment (ADmad) [inline src/Model/Resource.php]:
```suggestion
    public function __construct(array $properties = [], array $options = [])
```

File: src/Plugin.php [hunk 22]
```diff
@@ -15,21 +15,21 @@ class Plugin extends BasePlugin
      *
```
Comment (ADmad) [inline src/Plugin.php:15]: This class needs to be renamed to `WebservicePlugin`.

File: src/Datasource/Connection.php (hunk 2 above)
Comment (JustinRuiter) [reply src/Datasource/Connection.php]: The value from $config['driver'] on line 48 is not guaranteed to be AbstractDriver, it is mixed, so it seems to me that this check is required, especially because the property is typed.

Comment (ADmad) [reply src/Datasource/Connection.php]: If a wrong type is assigned to the property then PHP itself will throw a TypeError, so we don't need to explicitly throw one.

File: phpunit.xml.dist (hunk 1 above)
Comment (JustinRuiter) [reply phpunit.xml.dist]: Good suggestion, changed it.

File: src/Datasource/Connection.php (hunk 2 above)
Comment (JustinRuiter) [reply src/Datasource/Connection.php]: Phpstan forces us to be explicit, it will complain if we don't check the type of $tempDriver and I think being explicit is better than ignoring this line for phpstan.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (JustinRuiter) [reply .github/workflows/ci.yml]: Agreed, changed, we should follow cake's versioning support.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (JustinRuiter) [reply .github/workflows/ci.yml]: Okay, sure, added that.

File: composer.json (hunk 5 above)
Comment (JustinRuiter) [reply composer.json]: Makes sense, updated it.

File: src/Datasource/Connection.php (hunk 6 above)
Comment (JustinRuiter) [reply src/Datasource/Connection.php:44]: It is required for the function configName() (on line 94), a function required to be implemented by Connectioninterface

Comment (JustinRuiter) [reply src/Datasource/Connection.php:45]: Nvm, I now see your suggestion. Implemented.

File: src/Datasource/Connection.php (hunk 7 above)
Comment (JustinRuiter) [reply src/Datasource/Connection.php:96]: I don't see the benefit of this. I would still need to store the $config variable in $_config in the constructor, which is as much as a hassle as just storing the config name.

File: src/Datasource/Connection.php (hunk 2 above)
Comment (ADmad) [reply src/Datasource/Connection.php]: You can use `assert()` instead to make phpstan.

Comment (JustinRuiter) [reply src/Datasource/Connection.php]: Implemented

File: src/Datasource/Connection.php (hunk 8 above)
Comment (JustinRuiter) [reply src/Datasource/Connection.php]: Right, rolled that back

File: src/Datasource/Marshaller.php (hunk 9 above)
Comment (JustinRuiter) [reply src/Datasource/Marshaller.php]: Not sure where this got changed, but checked the cake changelogs and found that the error message was renamed to validate. Restored the code to the version in master and updated it to reflect that change.

File: src/Datasource/Query.php (hunk 10 above)
Comment (JustinRuiter) [reply src/Datasource/Query.php]: Seems like a nice, user friendly suggestion, added.

File: src/Datasource/Query.php (hunk 11 above)
Comment (JustinRuiter) [reply src/Datasource/Query.php]: Missed this one

File: src/Datasource/Query.php (hunk 12 above)
Comment (JustinRuiter) [reply src/Datasource/Query.php:391]: Me, psalm and phpstorm don't agree with that. Anyway, there are now no functions that return $this anymore anyway. They now all return self, because psalm presribes that.

File: src/Datasource/Query.php (hunk 13 above)
Comment (JustinRuiter) [reply src/Datasource/Query.php]: Moved

File: src/Datasource/Query.php (hunk 14 above)
Comment (JustinRuiter) [reply src/Datasource/Query.php]: Again, this is not good practice. I've changed the return type to what the interface requires. Our implementation should be in line with what the interface wants. Otherwise, there is not much use in implementing it.

File: src/Model/Endpoint.php (hunk 15 above)
Comment (JustinRuiter) [reply src/Model/Endpoint.php]: Reverting this change breaks EndpointLocator at line 96, where we then cannot be sure that we actually instantiate an endpoint. Do you have a suggestion to solve this in another way?

File: src/Model/Endpoint.php (hunk 16 above)
Comment (JustinRuiter) [reply src/Model/Endpoint.php]: Right, changed that

File: src/Model/Endpoint.php (hunk 17 above)
Comment (JustinRuiter) [reply src/Model/Endpoint.php]: Updated

File: src/Model/Endpoint.php (hunk 18 above)
Comment (JustinRuiter) [reply src/Model/Endpoint.php:674]: Right, will change that. I asumed (but that turns out to be wrong) that the first PR fixed most of these issues.

File: src/Model/EndpointLocator.php (hunk 19 above)
Comment (JustinRuiter) [reply src/Model/EndpointLocator.php]: The interface doesn't doesn't require that, but it makes sense indeed. Changed.

File: src/Model/EndpointLocator.php (hunk 20 above)
Comment (JustinRuiter) [reply src/Model/EndpointLocator.php]: Agreed, updated

File: src/Model/Resource.php (hunk 21 above)
Comment (JustinRuiter) [reply src/Model/Resource.php]: See the comment of Endpoint, I would like to ask you the same question here.

File: src/Plugin.php (hunk 22 above)
Comment (JustinRuiter) [reply src/Plugin.php:15]: Okay, changed it, but why? Doesn't this class always need to be called Plugin?

File: src/Datasource/Query.php (hunk 14 above)
Comment (ADmad) [reply src/Datasource/Query.php]: Cake's convention is to skip return type for all methods that return `$this` so please just follow that.

File: src/Model/Endpoint.php (hunk 15 above)
Comment (ADmad) [reply src/Model/Endpoint.php]: I don't see how it breaks the `EndpointLocator`. The core's `Table` class isn't `final` and there are new issues in it's `TableLocator`.

File: src/Plugin.php (hunk 22 above)
Comment (ADmad) [reply src/Plugin.php:15]: Cake 5's convention is that the plugins classes should be named as `{PluginName}Plugin`.

Comment (JustinRuiter) [reply src/Plugin.php:15]: Okay, didn't know that. Thanks for the heads up!

File: src/Model/Endpoint.php (hunk 15 above)
Comment (JustinRuiter) [reply src/Model/Endpoint.php]: Interesting, I'll look into how it's done in core, and update the code to work like that instead.

File: src/Datasource/Query.php (hunk 14 above)
Comment (JustinRuiter) [reply src/Datasource/Query.php]: Okay, but then we would create psalm warnings, should I just ignore those then? Or is there a different way in which these warnings should be fixed or ignored?

File: tests/test_app/src/Model/Endpoint/TestEndpoint.php [hunk 23]
```diff
@@ -32,7 +32,7 @@ public function validationDefault(Validator $validator): Validator
      *
      * @return true
      */
-    public function findExamples()
+    public function findExamples(): true
```
Comment (ADmad) [inline tests/test_app/src/Model/Endpoint/TestEndpoint.php]:
```suggestion
    public function findExamples(): bool
```
`true` return type is available since PHP 8.2 only

Comment (JustinRuiter) [reply tests/test_app/src/Model/Endpoint/TestEndpoint.php]: Ahh, I was about to install php 8.1 to test this locally. Will still install it though ;). Thanks!

File: src/Model/Endpoint.php (hunk 15 above)
Comment (JustinRuiter) [reply src/Model/Endpoint.php]: Solved this by adding a psalm constraint, as dictated by psalm. This solution thus only works if any project dependent on Muffin also uses psalm.

File: src/Model/Resource.php (hunk 21 above)
Comment (JustinRuiter) [reply src/Model/Resource.php]: Solved this by adding a psalm constraint, as dictated by psalm. This solution thus only works if any project dependent on Muffin also uses psalm.

File: src/Datasource/Connection.php (hunk 7 above)
Comment (JustinRuiter) [reply src/Datasource/Connection.php:96]: @ADmad if you agree, please resolve the conversation, or propose a different solution.

File: src/Datasource/Connection.php (hunk 6 above)
Comment (JustinRuiter) [reply src/Datasource/Connection.php:44]: Can be closed if the comment below is closed.

File: src/Model/Endpoint.php (hunk 18 above)
Comment (JustinRuiter) [reply src/Model/Endpoint.php:674]: Seeing that the findList method in core returns a SelectQuery. Do you want me to implement the different query types core has here as well? Or just the way this method works? @ADmad

File: src/Datasource/Query.php [hunk 24]
```diff
@@ -99,12 +161,251 @@ public function __construct(WebserviceInterface $webservice, Endpoint $endpoint)
...
+            return $this->_results;
+        }
+
+        /** @psalm-suppress InternalMethod Could not find a better way apart from implementing it as a custom class **/
+        $results = $this->_cache?->fetch($this);
+        if ($results === null) {
+            $res = $this->_execute();
+
+            if (!is_iterable($res)) {
+                return new ResultSet([], 0);
```
Comment (ADmad) [inline src/Datasource/Query.php]: Why aren't we setting `$this->_results` here?

File: src/Datasource/Query.php [hunk 25]
```diff
@@ -99,12 +161,251 @@ public function __construct(WebserviceInterface $webservice, Endpoint $endpoint)
...
+        }
+
+        if (is_array($fields) && empty($fields)) {
+            return $this;
+        }
+
+        $this->_parts['order'] ??= new OrderByExpression();
+        $this->_conjugate('order', $fields, '', []);
+
+        return $this;
```
Comment (ADmad) [inline src/Datasource/Query.php]:
```suggestion
        return $this->order($fields, $overwrite);
```

File: src/Datasource/Query.php [hunk 26]
```diff
@@ -99,12 +161,251 @@ public function __construct(WebserviceInterface $webservice, Endpoint $endpoint)
...
+     * Helper function used to build conditions by composing QueryExpression objects.
+     *
+     * @param string $part Name of the query part to append the new part to
+     * @param \Cake\Database\ExpressionInterface|\Closure|array|string|null $append Expression or builder function to append.
+     *   to append.
+     * @param string $conjunction type of conjunction to be used to operate part
+     * @param array<string, string> $types Associative array of type names used to bind values to query
+     * @return void
+     */
+    protected function _conjugate(
```
Comment (ADmad) [inline src/Datasource/Query.php]: Is this method actually called from somewhere? This and other methods like `expr()`, `newExpr()` which deal with db expressions make no sense in context of this query class. They should be dropped.

File: src/Datasource/Query.php [hunk 27]
```diff
@@ -272,18 +570,20 @@ public function aliasField(string $field, ?string $alias = null): array
...
-        if ($conditions === null) {
-            return $this->clause('where');
+    public function where(
+        Closure|array|string|null $conditions = null,
+        array $types = [],
+        bool $overwrite = false
+    ): Query {
+        if ($overwrite) {
+            $this->_parts['where'] = $conditions;
         }
```
Comment (ADmad) [inline src/Datasource/Query.php]:
```suggestion
```
`$overwrite` is already taken in account in the line below.

File: src/Model/Endpoint.php (hunk 18 above)
Comment (ADmad) [reply src/Model/Endpoint.php:674]: Yes eventually do I want to use separate query classes for this plugin too. That would allow us to clean up the types for the `Query::$_results` property. Currently it can be of multiple different types which makes the code messy as we have to add type checks wherever it's used, as evident from your changes.

File: src/Datasource/Query.php (hunk 27 above)
Comment (JustinRuiter) [reply src/Datasource/Query.php]: Right

File: src/Datasource/Query.php (hunk 24 above)
Comment (JustinRuiter) [reply src/Datasource/Query.php]: Updated

File: src/Datasource/Query.php (hunk 26 above)
Comment (JustinRuiter) [reply src/Datasource/Query.php]: Good catch, not sure why it was in there even.

**Comment by JustinRuiter** [comment]
I don't really have the time or expertise to implement the seperate query classes from core. Could this be picked up in a seperate issue?

**Comment by ADmad** [comment]
> Could this be picked up in a seperate issue?

Yup
